### PR TITLE
python312Packages.wandb: 0.16.0 -> 0.18.5

### LINKED ISSUE #349691: Build failure: python311Packages.wandb
### Describe the bug
Aider chat breaks home-manager build on ubuntu, seems to be since https://github.com/NixOS/nixpkgs/commit/1a7ad9eabbb44da5063195ffa2f6bf2056902623

aider-chat is only present on nixpkgs-unstable, so maybe just a deal with it. but want it to work so thought I'd report

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Things done

Changelog: https://github.com/wandb/wandb/raw/v0.18.5/CHANGELOG.md

cc @samuela

fixes https://github.com/NixOS/nixpkgs/issues/349691

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by DerDennisOP** [comment]
nice, wandb update :)

**Comment by GaetanLepage** [comment]
> nice, wandb update :)

I am still battling with a few tests but the package seems to be working nicely overall.

**Comment by GaetanLepage** [comment]
I got it to build on linux, but the rust binary fails to compile on darwin:
```
error: linking with `/nix/store/22a6p8311lg3z4ksrjbj018i7z3vgkva-clang-wrapper-16.0.6/bin/cc` failed: exit status: 1
  |
  = note: env -u IPHONEOS_DEPLOYMENT_TARGET -u TVOS_DEPLOYMENT_TARGET -u XROS_DEPLOYMENT_TARGET LC_ALL="C" PATH="/nix/store/rgd8c9c4d3j1aqdd8mbvq1ll2kcpzmrd-rustc-1.81.0/lib/rustlib>
  = note: ld: library not found for -lIOReport
          clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by samuela** [comment]
I'm seeing the aarch64-darwin build fail for a different reason:
```
❯ nix-build -A python3Packages.wandb   
...
       error: Package ‘gpu-stats-0.2.0’ in /Users/samuelainsworth/dev/nixpkgs/pkgs/development/python-modules/wandb/default.nix:88 is marked as broken, refusing to evaluate.
...
```
@GaetanLepage is there something different about your setup?

FWIW I'm happy to merge without darwin support since building on linux is an improvement over the current status quo.

**Comment by GaetanLepage** [comment]
I am about to push the following which I think is better:
```diff
index 3de94a2a99f1..676c46295095 100644
--- a/pkgs/development/python-modules/wandb/default.nix
+++ b/pkgs/development/python-modules/wandb/default.nix
@@ -162,22 +162,34 @@ buildPythonPackage rec {
     })
   ];

+  env = {
+    # https://github.com/wandb/wandb/blob/v0.18.5/hatch_build.py#L37-L47
+
+    # Prevent the install script to try building and embedding the `gpu_stats` binary in the wheel.
+    # It will be patched within the `wandb-core` source code.
+    WANDB_BUILD_SKIP_GPU_STATS = true;
+
+    # Prevent the install script to try building and embedding the `wandb-core` binary in the wheel.
+    # It will be patched within the `wandb` source code.
+    WANDB_BUILD_UNIVERSAL = true;
+  };
+
   # We manually compile those two dependencies and inject their path to bypass their build scripts
   postPatch = ''
-    substituteInPlace hatch_build.py \
-      --replace-fail \
-        'output = pathlib.Path("wandb", "bin", "wandb-core")' \
-        'return ["${lib.getExe wandb-core}"]' \
-      --replace-fail \
-        'output = pathlib.Path("wandb", "bin", "gpu_stats")' \
-        'return ["${lib.getExe gpu-stats}"]'
-
     substituteInPlace wandb/util.py \
       --replace-fail \
         'bin_path = pathlib.Path(__file__).parent / "bin" / "wandb-core"' \
         'bin_path = pathlib.Path("${lib.getExe wandb-core}")'
   '';

+  # substituteInPlace hatch_build.py \
+  #   --replace-fail \
+  #     'output = pathlib.Path("wandb", "bin", "wandb-core")' \
+  #     'return ["${lib.getExe wandb-core}"]' \
+  #   --replace-fail \
+  #     'output = pathlib.Path("wandb", "bin", "gpu_stats")' \
+  #     'return ["${lib.getExe gpu-stats}"]'
+
   build-system = [
     hatchling
   ];
```

Indeed, currently I am still "embedding" the `gpu_stats` and `wandb-core` binaries in the wheel (which is not needed).

**Comment by GaetanLepage** [comment]
Ok @samuela, this is good to go according to me. Please, have a look at it to spot eventual mistakes.

- The failures are not regressions. Most of them are caused by failing `python3Packages.opentelemetry-proto`.
- I was able to build it fine on `x86_64-darwin`. Some tests (matplotlib/GUI related) were completely crashing the interpreter. A few others were failing. I disabled them.
- The `gpu_stats` rust binary still fails to build so I marked the package as broken on this platform. Nevertheless I was able to build it just fine from `staging-next`, so we should be able to retry once it has landed on `master`.

File: pkgs/development/python-modules/kubernetes-asyncio/default.nix
Comment (samuela) [inline pkgs/development/python-modules/kubernetes-asyncio/default.nix:60]: It looks like this is deprecated/renamed:
https://github.com/NixOS/nixpkgs/blob/7783a48f3858b2783be21167118735ae8407e064/pkgs/stdenv/generic/make-derivation.nix#L130-L137
but I'm not able to find any docs pointing to why or what to use instead.

Perhaps @NixOS/darwin-maintainers might be able to help us?

File: pkgs/development/python-modules/wandb/default.nix
Comment (samuela) [inline pkgs/development/python-modules/wandb/default.nix:140]: same note here re this flag being deprecated/removed

File: pkgs/development/python-modules/wandb/default.nix
Comment (samuela) [inline pkgs/development/python-modules/wandb/default.nix:265]:
```
❯ ns -p python312
[nix-shell:~]$ which python3
/nix/store/ybnf7k6i9p244bbhsbxizqk65z58cwyr-python3-3.12.6/bin/python3
```
perhaps I'm missing something here?

**Review by samuela — COMMENTED** [review]
This is great -- thanks so much @GaetanLepage! I'm a little confused how `__darwinAllowLocalNetworking` is intended to be used, but o/w I think this looks good.

File: pkgs/development/python-modules/wandb/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/wandb/default.nix:265]: I know, this is weird. But I confirm that the test fails:
```
_______________________ test_get_entrypoint[wandb_core] ________________________
[gw0] linux -- Python 3.12.6 /nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/bin/python3.12

    @pytest.mark.skipif(
        platform.system() == "Windows",
        reason="python exec name is different on windows",
    )
    def test_get_entrypoint():
        dir = tempfile.TemporaryDirectory().name
        job_source = "artifact"
        builder = _configure_job_builder_for_partial(dir, job_source)
    
        metadata = {"python": "3.9.11", "codePathLocal": "main.py", "_partial": "v0"}
    
        program_relpath = builder._get_program_relpath(job_source, metadata)
        entrypoint = builder._get_entrypoint(program_relpath, metadata)
>       assert entrypoint == ["python3", "main.py"]
E       AssertionError: assert ['python3.12', 'main.py'] == ['python3', 'main.py']
E         
E         At index 0 diff: 'python3.12' != 'python3'
E         
E         Full diff:
E           [
E         -     'python3',
E         +     'python3.12',
E         ?             +++
E               'main.py',
E           ]

/build/source/tests/unit_tests/test_launch/test_create_job.py:126: AssertionError
```

File: pkgs/development/python-modules/kubernetes-asyncio/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/kubernetes-asyncio/default.nix:60]: Apparently, it will just be removed from the derivation arguments and we can still use it without any problem.

**Review by drupol — APPROVED** [review]

**Comment by GaetanLepage** [comment]
@samuela do you have any further concern ?
I would like to have this merged soon if possible as it blocks other PRs that I have prepared since then.

**Comment by drupol** [comment]
I took the decision to merge it so we can move on, let's iterate on this later if @samuela has some other concerns.

**Pull request merged**
